website/project/views/comment.py

```diff
@@ -15,104 +15,214 @@
+            collect_discussion(comment, users=users)
     return users
 
 
-@must_be_contributor_or_public
-def comment_discussion(**kwargs):
+def comment_discussion(comments, node, anonymous=False, widget=False):
 
-    node = kwargs['node'] or kwargs['project']
-    auth = kwargs['auth']
-    users = collect_discussion(node)
-    anonymous = has_anonymous_link(node, auth)
-    # Sort users by comment frequency
-    # TODO: Allow sorting by recency, combination of frequency and recency
-    sorted_users = sorted(
+    users = collections.defaultdict(list)
+    for comment in comments:
+        if not (comment.is_deleted or comment.is_hidden):
+            users[comment.user].append(comment)
+        if not widget:
+            collect_discussion(comment, users=users)
+
+    sorted_users_frequency = sorted(
         users.keys(),
         key=lambda item: len(users[item]),
         reverse=True,
     )
 
+    def get_recency(item):
+        most_recent = users[item][0].date_created
+        for comment in users[item][1:]:
+            if comment.date_created > most_recent:
+                most_recent = comment.date_created
+        return most_recent
+
+    sorted_users_recency = sorted(
+        users.keys(),
+        key=lambda item: get_recency(item),
+        reverse=True,
+    )
+
     return {
-        'discussion': [
-            {
-                'id': privacy_info_handle(user._id, anonymous),
-                'url': privacy_info_handle(user.url, anonymous),
-                'fullname': privacy_info_handle(user.fullname, anonymous, name=True),
-                'isContributor': node.is_contributor(user),
-                'gravatarUrl': privacy_info_handle(
-                    gravatar(
-                        user, use_ssl=True, size=settings.GRAVATAR_SIZE_DISCUSSION,
-                    ),
-                    anonymous
-                ),
-
-            }
-            for user in sorted_users
+        'discussionByFrequency': [
+            serialize_discussion(node, user, len(users[user]), anonymous)
+            for user in sorted_users_frequency
+        ],
+        'discussionByRecency': [
+            serialize_discussion(node, user, len(users[user]), anonymous)
+            for user in sorted_users_recency
         ]
     }
 
 
+def serialize_discussion(node, user, num, anonymous=False):
+    return {
+        'id': privacy_info_handle(user._id, anonymous),
+        'url': privacy_info_handle(user.url, anonymous),
+        'fullname': privacy_info_handle(user.fullname, anonymous, name=True),
+        'isContributor': node.is_contributor(user),
+        'gravatarUrl': privacy_info_handle(
+            gravatar(
+                user, use_ssl=True, size=settings.GRAVATAR_SIZE_DISCUSSION,
+            ),
+            anonymous
+        ),
+        'numOfComments': num
+    }
+
+
 def serialize_comment(comment, auth, anonymous=False):
+    from website.addons.wiki.model import NodeWikiPage
```

Comment:
Is this necessary? Does importing at the top of the file cause a circular import? If it does, can this function be moved somewhere to avoid the circular import?